lib/vnet/vnet.go

```diff
@@ -65,9 +76,41 @@ func (c *Config) CheckAndSetDefaults() error {
 	if c.IPv6Prefix.Len() != 16 || c.IPv6Prefix.AsSlice()[0] != 0xfd {
 		return trace.BadParameter("IPv6Prefix must be an IPv6 ULA address")
 	}
+	if c.TCPHandlerResolver == nil {
+		return trace.BadParameter("TCPHandlerResolver is required")
+	}
 	return nil
 }
 
+// TCPHandlerResolver describes a type that can resolve a fully-qualified domain name to a TCP handler that
+// should handle all future TCP connections to that FDQN.
+//
+// Implementations beware - an FQDN always ends with a '.'.
+type TCPHandlerResolver interface {
+	// ResolveTCPHandler decides if `fqdn` should match a TCP handler.
+	//
+	// If [fqdn] matches it must return a TCPHandler for future connections to any assigned IPs.
+	//
+	// If [fqdn] does not match an app it must return with match == false && err == nil, in this case the DNS
+	// request will be forwarded to upstream nameservers. Only return a non-nil error for truly unexpected
+	// errors that should cause a DNS request to fail.
+	ResolveTCPHandler(ctx context.Context, fqdn string) (handler TCPHandler, match bool, err error)
+}
+
+// TCPHandler defines the behavior for handling TCP connections from VNet.
+//
+// Implementations should attempt to dial the target application and return any errors before calling
+// [connector] to complete the TCP handshake and get the TCP conn. This is so that clients will see that the
+// TCP connection was refused, instead of seeing a successful TCP dial that is immediately closed.
+type TCPHandler interface {
+	HandleTCPConnector(ctx context.Context, connector func() (net.Conn, error)) error
+}
+
+// UDPHandler defines the behavior for handling UDP connections from VNet.
+type UDPHandler interface {
+	HandleUDP(context.Context, net.Conn) error
+}
```

Comment:
Are these always going to be DNS? If so, it might be worth clarifying that in the name (or at least the docstring).
```suggestion
// UDPHandler defines the behavior for handling UDP connections from VNet.
type DNSOverUDPHandler interface {
	HandleDNSOverUDP(context.Context, net.Conn) error
}
```

Reply:
the vnet manager is agnostic to the application layer protocol, the UDP handler could do anything, DNS or whatever. but yeah you're right, in practice, at least for now, it's only DNS